### PR TITLE
Embedding parameters

### PR DESCRIPTION
**Pull request by YannFollet**
### --embd-normalize $integer$
| $integer$ | description         | formula |
|-----------|---------------------|---------|
| $-1$      | none                |
| $0$       | max absolute int16  | $\Large{{32760 * x_i} \over\max \lvert x_i\rvert}$
| $1$       | taxicab             | $\Large{x_i \over\sum \lvert x_i\rvert}$
| $2$       | euclidean (default) | $\Large{x_i \over\sqrt{\sum x_i^2}}$
| $>2$      | p-norm              | $\Large{x_i \over\sqrt[p]{\sum \lvert x_i\rvert^p}}$

### --embd-output-format $'string'$
| $'string'$ | description                  |  |
|------------|------------------------------|--|
| ''         | same as before               | (default)
| 'array'    | single embeddings            | $[[x_1,...,x_n]]$
|            | multiple embeddings          | $[_0[x_1,...,x_n],_1[x_1,...,x_n],...,_{n-1}[x_1,...,x_n]]$
| 'json'     | openai style                 |
| 'json+'    | add cosine similarity matrix |

### --embd-separator $"string"$
| $"string"$   | |
|--------------|-|
| "\n"         | (default)
| "<#embSep#>" | for exemple 
| "<#sep#>"    | other exemple 

### Unix-based systems (Linux, macOS, etc.):

```bash
./embedding -p 'Castle<#sep#>Stronghold<#sep#>Dog<#sep#>Cat' --embd-separator '<#sep#>' --embd-normalize 2  --embd-output-format '' -m './path/to/model.gguf' --n-gpu-layers 99 --log-disable 2>/dev/null
```

### Windows:

```powershell
embedding.exe -p 'Castle<#sep#>Stronghold<#sep#>Dog<#sep#>Cat' --embd-separator '<#sep#>' --embd-normalize 2  --embd-output-format '' -m './path/to/model.gguf' --n-gpu-layers 99 --log-disable 2>/dev/null
```

### CODE REVIEW

**Review by ggerganov — APPROVED** [review]

**Comment by mofosyne** [comment]
@YannFollet is this PR all good? Have a quick lookover. Had to do some merging to keep this in sync.

Marking as merging soon and will revisit later to see if anyone want to still comment about it.

**Comment by mofosyne** [comment]
@YannFollet this PR has gone out of sync. Can you synchronize this with the main branch? It appears there is no other objection and GG is happy with it. But it's out of sync so cannot easily merge.

**Comment by YannFollet** [comment]
@mofosyne Sorry haven't seen your message, I have done the merge to the master

File: examples/embedding/embedding.cpp [hunk 1]
```diff
@@ -182,29 +182,77 @@ int main(int argc, char ** argv) {
...
-        for (int i = 0; i < (n_prompts > 1 ? std::min(16, n_embd) : n_embd); i++) {
-            fprintf(stdout, "%9.6f ", emb[j * n_embd + i]);
-        }
-        fprintf(stdout, "\n");
-    }
+    batch_decode(ctx, batch, out, s, n_embd, params.embd_normalize);
 
-    // print cosine similarity matrix
-    if (n_prompts > 1) {
+    if (params.embd_out=="") {
```
Comment (ggerganov) [inline examples/embedding/embedding.cpp:187]:
```suggestion
    if (params.embd_out.empty()) {
```

File: common/common.h [hunk 2]
```diff
@@ -153,6 +153,9 @@ struct gpt_params {
     bool prompt_cache_ro   = false; // open the prompt cache read-only and do not update it
 
     bool embedding         = false; // get only sentence embedding
+    int32_t embd_normalize = 2;     // normalisation for embendings (-1=none, 0=max absolute, 1=taxicab, 2=euclidean, >2=p-norm)
+    std::string embd_out   = "";    // empty = default, "array" = [] or [[],[]...], "json" = openai style, "json+" = same "json" + cosine similarity matrix
+    std::string embd_sep   = "\n";  // separator of embendings
```
Comment (ggerganov) [inline common/common.h:158]: These group of parameters should be moved below in their own section, similar to what we do for other examples:

```c++
    // embedding
    int32_t embd_normalize = 2;     // normalisation for embedings (-1=none, 0=max absolute, 1=taxicab, 2=euclidean, >2=p-norm)

    std::string embd_out   = "";    // empty = default, "array" = [] or [[],[]...], "json" = openai style, "json+" = same "json" + cosine similarity matrix
    std::string embd_sep   = "\n";  // separator of embedings
```

File: common/common.cpp [hunk 3]
```diff
@@ -922,6 +714,21 @@ bool gpt_params_find_arg(int argc, char ** argv, const std::string & arg, gpt_pa
...
+    if (arg == "--embd-output-format") {
+        CHECK_ARG
+        params.embd_out = argv[i];
+        return true;
+    }
+    if (arg == "--embd-separator") {
+        CHECK_ARG
+        params.embd_sep = argv[i];
+        return true;
+    }
```
Comment (ggerganov) [inline common/common.cpp:731]: Expand `gpt_params_print_usage` for these parameters

File: common/common.h [hunk 4]
```diff
@@ -377,7 +380,7 @@ void llama_kv_cache_dump_view_seqs(const llama_kv_cache_view & view, int row_siz
 // Embedding utils
 //
 
-void llama_embd_normalize(const float * inp, float * out, int n);
+void llama_embd_normalize(const float * inp, float * out, int n, int embd_norm  = 2);
```
Comment (ggerganov) [inline common/common.h:383]:
```suggestion
void llama_embd_normalize(const float * inp, float * out, int n, int embd_norm = 2);
```

File: examples/embedding/embedding.cpp [hunk 5]
```diff
@@ -182,29 +182,77 @@ int main(int argc, char ** argv) {
...
-            for (int j = 0; j < n_prompts; j++) {
-                float sim = llama_embd_similarity_cos(emb + i * n_embd, emb + j * n_embd, n_embd);
-                fprintf(stdout, "%6.2f ", sim);
+        for (int j = 0; j < n_prompts; j++) {
+            fprintf(stdout, "embedding %d: ", j);
+            for (int i = 0; i < (n_prompts > 1 ? std::min(16, n_embd) : n_embd); i++) {
+                if (params.embd_normalize==0)
+                    fprintf(stdout, "%6.0f ", emb[j * n_embd + i]);
+                else
+                    fprintf(stdout, "%9.6f ", emb[j * n_embd + i]);
```
Comment (ggerganov) [inline examples/embedding/embedding.cpp:196]:
```suggestion
                if (params.embd_normalize == 0) {
                    fprintf(stdout, "%6.0f ", emb[j * n_embd + i]);
                } else {
                    fprintf(stdout, "%9.6f ", emb[j * n_embd + i]);
                }
```

File: examples/embedding/embedding.cpp [hunk 6]
```diff
@@ -182,29 +182,77 @@ int main(int argc, char ** argv) {
...
+            for (int i = 0; i < n_prompts; i++) {
+                fprintf(stdout, "%6.6s ", prompts[i].c_str());
             }
             fprintf(stdout, "\n");
+            for (int i = 0; i < n_prompts; i++) {
+                for (int j = 0; j < n_prompts; j++) {
+                    float sim = llama_embd_similarity_cos(emb + i * n_embd, emb + j * n_embd, n_embd);
+                    fprintf(stdout, "%6.2f ", sim);
+                }
+                fprintf(stdout, "%1.10s",prompts[i].c_str());
```
Comment (ggerganov) [inline examples/embedding/embedding.cpp:214]:
```suggestion
                fprintf(stdout, "%1.10s", prompts[i].c_str());
```

File: examples/embedding/embedding.cpp [hunk 7]
```diff
@@ -182,29 +182,77 @@ int main(int argc, char ** argv) {
...
+                    if (j < n_prompts) fprintf(stdout, ", "); else break;
+                }
+                fprintf(stdout, " ]");
+                i++;
+                if (i < n_prompts) fprintf(stdout, ",\n"); else break;
+            }
+            fprintf(stdout, "\n  ]");
+        }
+
+        if (notArray) fprintf(stdout, "\n}\n");
```
Comment (ggerganov) [inline examples/embedding/embedding.cpp:255]:
```suggestion
    if (params.embd_out == "json" || params.embd_out == "json+" || params.embd_out == "array") {
        const bool notArray = params.embd_out != "array";

        fprintf(stdout, notArray ? "{\n  \"object\": \"list\",\n  \"data\": [\n":"[");
        for (int j = 0;;) { // at least one iteration (one prompt)
            if (notArray) fprintf(stdout, "    {\n      \"object\": \"embedding\",\n      \"index\": %d,\n      \"embedding\": ",j);
            fprintf(stdout, "[");
            for (int i = 0;;) { // at least one iteration (n_embd > 0)
                fprintf(stdout, params.embd_normalize == 0 ? "%1.0f" : "%1.7f", emb[j * n_embd + i]);
                i++;
                if (i < n_embd) fprintf(stdout, ","); else break;
            }
            fprintf(stdout, notArray ? "]\n    }" : "]");
            j++;
            if (j < n_prompts) fprintf(stdout, notArray ? ",\n" : ","); else break;
        }
        fprintf(stdout, notArray ? "\n  ]" : "]\n");

        if (params.embd_out == "json+" && n_prompts > 1) {
            fprintf(stdout, ",\n  \"cosineSimilarity\": [\n");
            for (int i = 0;;) { // at least two iteration (n_prompts > 1)
                fprintf(stdout, "    [");
                for (int j = 0;;) { // at least two iteration (n_prompts > 1)
                    float sim = llama_embd_similarity_cos(emb + i * n_embd, emb + j * n_embd, n_embd);
                    fprintf(stdout, "%6.2f", sim);
                    j++;
                    if (j < n_prompts) fprintf(stdout, ", "); else break;
                }
                fprintf(stdout, " ]");
                i++;
                if (i < n_prompts) fprintf(stdout, ",\n"); else break;
            }
            fprintf(stdout, "\n  ]");
        }

        if (notArray) fprintf(stdout, "\n}\n");
```

**Pull request merged**